ghcide/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -86,7 +125,43 @@ instance NFData ShowDiagnostic where
 --   along with the related source location so that we can display the error
 --   on either the console or in the IDE at the right source location.
 --
-type FileDiagnostic = (NormalizedFilePath, ShowDiagnostic, Diagnostic)
+data FileDiagnostic = FileDiagnostic
+  { fdFilePath :: NormalizedFilePath
+  , fdShouldShowDiagnostic :: ShowDiagnostic
+  , fdLspDiagnostic :: Diagnostic
+  , fdStructuredMessage :: StructuredMessage
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+instance NFData FileDiagnostic
+
+data StructuredMessage
+  = NoStructuredMessage
+  | SomeStructuredMessage (MsgEnvelope GhcMessage)
+  deriving (Generic)
+
+instance Show StructuredMessage where
+  show NoStructuredMessage = "NoStructuredMessage"
+  show SomeStructuredMessage {} = "SomeStructuredMessage"
+
+instance Eq StructuredMessage where
+  (==) NoStructuredMessage NoStructuredMessage = True
+  (==) SomeStructuredMessage {} SomeStructuredMessage {} = True
+  (==) _ _ = False
```

Comment:
Are these instances safe? Do we need them? Worth a comment.

Reply:
Documented in 003c15c1 , further discussion in https://github.com/haskell/haskell-language-server/pull/4311#discussion_r1641871002